common/src/main/scala/org/neo4j/spark/reader/BasePartitionReader.scala

```diff
@@ -81,37 +84,63 @@ abstract class BasePartitionReader(
   @volatile
   private var error: Boolean = false
 
+  private val retries = new AtomicInteger(options.transactionSettings.retries)
+
   @throws(classOf[IOException])
   def next: Boolean =
     try {
-      if (result == null) {
-        session = driverCache.getOrCreate().session(options.session.toNeo4jSession())
-        transaction = session.beginTransaction(options.toNeo4jTransactionConfig)
-
-        val queryText = query()
-        val queryParams = queryParameters
-
-        logInfo(s"Running the following query on Neo4j: $queryText")
-        logDebug(s"with parameters $queryParams")
-
-        result = transaction.run(queryText, Values.value(queryParams))
-          .asScala
-      }
-
-      if (result.hasNext) {
-        nextRow = mappingService.convert(result.next(), schema)
-        true
-      } else {
-        false
-      }
+      nextHandler()
     } catch {
-      case t: Throwable => {
-        error = true
-        logInfo("Error while invoking next:", t)
-        throw new IOException(t)
-      }
+      case t: Throwable =>
+        if (options.transactionSettings.shouldFailOn(t)) {
+          error = true
+          logError("Error while invoking next due to explicitly configured failure condition:", t)
+          throw new IOException(t)
+        }
+
+        if (Neo4jUtil.isRetryableException(t) && retries.get() > 0) {
+          val currentRetry = retries.decrementAndGet
```

Comment:
This is not very atomic operation and it can increase the number of retries if used in multiple threads, but I don't think it's a problem.